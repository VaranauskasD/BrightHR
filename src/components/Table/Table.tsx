import React, { useEffect, useState } from 'react'
import styled from 'styled-components'

import { AiOutlineArrowDown, AiOutlineArrowUp } from 'react-icons/Ai'

import { DataFile } from '../types'
import { type } from 'os'

export interface TableProps {
  data: DataFile[]
}

const StyledTable = styled.table`
  display: flex;
  flex-direction: column;
  padding: 8px;
  padding-top: 24px;
  width: 100%;
`

const StyledCaption = styled.caption`
  font-size: 24px;
  text-align: left;
`

const StyledTableHead = styled.thead`
  height: 100%;
  background: ${(props) => `${props.theme.colors.tertiary}`};
  margin: 4px 0;
`

const StyledTableHeadRow = styled.tr`
  display: flex;
  justify-content: space-between;
`

const StyledTableBodyRow = styled.tr<{ $key: number }>`
  display: flex;
  justify-content: space-between;
  background: ${(props) =>
    props.$key % 2
      ? `${props.theme.colors.grey}`
      : `${props.theme.colors.secondary}`};
`

const StyledTableHeader = styled.th`
  width: 100%;
  min-height: 32px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`

const TableButton = styled.button`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  background: none;
  border: none;
  cursor: pointer;
`

const StyledTableBody = styled.tbody`
  height: 100%;
`

const StyledTableCell = styled.td`
  width: 100%;
  min-height: 32px;
  padding: 8px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`

const StyledArrowUpIcon = styled(AiOutlineArrowUp)`
  margin: 0 4px;
`

const StyledArrowDownIcon = styled(AiOutlineArrowDown)`
  margin: 0 4px;
`

export const Table = (props: TableProps) => {
  const [tableData, setTableData] = useState<DataFile[]>([])
  const [sortKeys, setSortKeys] = useState<number[]>([1, 0, 0, 0])
  const [nameAscending, setNameAscending] = useState<boolean>(true)
  const [typeAscending, setTypeAscending] = useState<boolean>(true)
  const [dateAscending, setDateAscending] = useState<boolean>(true)
  const [sizeAscending, setSizeAscending] = useState<boolean>(true)

  useEffect(() => {
    sortByName(props.data, true)
    setTableData(props.data)
  }, [])

  const sortByName = (data: DataFile[], ascending: boolean) => {
    const sortData = data

    ascending
      ? sortData.sort((file_one, file_two) =>
          file_one.name.localeCompare(file_two.name)
        )
      : sortData.sort((file_one, file_two) =>
          file_two.name.localeCompare(file_one.name)
        )

    setTableData(sortData)
    setNameAscending(!nameAscending)

    const keys = Array(4).fill(0)
    keys[0] = 1
    setSortKeys(keys)
  }

  const sortByType = (data: DataFile[], ascending: boolean) => {
    const sortData = data

    ascending
      ? sortData.sort((file_one, file_two) => {
          if (file_one.type === 'folder' && file_two.type !== 'folder')
            return -1
          if (file_two.type === 'folder' && file_one.type !== 'folder') return 1
          return file_one.type.localeCompare(file_two.type)
        })
      : sortData.sort((file_one, file_two) => {
          if (file_one.type === 'folder' && file_two.type !== 'folder') return 1
          if (file_two.type === 'folder' && file_one.type !== 'folder')
            return -1
          return file_two.type.localeCompare(file_one.type)
        })

    setTableData(data)
    setTypeAscending(!typeAscending)

    const keys = Array(4).fill(0)
    keys[1] = 1
    setSortKeys(keys)
  }

  const sortByDate = (data: DataFile[], ascending: boolean) => {
    const sortData = data

    ascending
      ? sortData.sort((file_one, file_two) => {
          if (!file_one.added) return 1
          if (!file_two.added) return -1
          return file_one.added.localeCompare(file_two.added)
        })
      : sortData.sort((file_one, file_two) => {
          if (!file_one.added) return 1
          if (!file_two.added) return -1
          return file_two.added.localeCompare(file_one.added)
        })

    setTableData(data)
    setDateAscending(!dateAscending)

    const keys = Array(4).fill(0)
    keys[2] = 1
    setSortKeys(keys)
  }

  const sortBySize = (data: DataFile[], ascending: boolean) => {
    const sortData = data

    ascending
      ? sortData.sort((file_one, file_two) => {
          if (!file_one.files) return 1
          if (!file_two.files) return -1
          return file_one.files.length - file_two.files.length
        })
      : sortData.sort((file_one, file_two) => {
          if (!file_one.files) return 1
          if (!file_two.files) return -1
          return file_two.files.length - file_one.files.length
        })

    setTableData(data)
    setSizeAscending(!sizeAscending)

    const keys = Array(4).fill(0)
    keys[3] = 1
    setSortKeys(keys)
  }

  const dataRow = (data: DataFile, key: number) => {
    return (
      <StyledTableBodyRow key={`${data.name}-${key}`} $key={key}>
        <StyledTableCell>
          {data.name ? data.name : 'Missing Name'}
        </StyledTableCell>
        <StyledTableCell>
          {data.type ? data.type : 'Missing Type'}
        </StyledTableCell>
        <StyledTableCell>
          {data.added ? data.added : 'Missing Date'}
        </StyledTableCell>
        <StyledTableCell>
          {data.files ? data.files.length : 'Missing size'}
        </StyledTableCell>
      </StyledTableBodyRow>
    )
  }

  return (
    <StyledTable aria-labelledby="caption">
      <StyledCaption id="caption">Files</StyledCaption>
      <StyledTableHead>
        <StyledTableHeadRow>
          <StyledTableHeader
            id="ColName"
            aria-sort={
              sortKeys[0]
                ? nameAscending
                  ? 'ascending'
                  : 'descending'
                : 'none'
            }
          >
            <TableButton
              type="button"
              id="ColNameSortButton"
              aria-label="Sort by name"
              onClick={() => sortByName(tableData, nameAscending)}
            >
              <span>Name</span>
              {sortKeys[0] ? (
                nameAscending ? (
                  <StyledArrowUpIcon aria-hidden="true" />
                ) : (
                  <StyledArrowDownIcon aria-hidden="true" />
                )
              ) : (
                <React.Fragment />
              )}
            </TableButton>
          </StyledTableHeader>
          <StyledTableHeader
            id="ColType"
            aria-sort={
              sortKeys[1]
                ? typeAscending
                  ? 'ascending'
                  : 'descending'
                : 'none'
            }
          >
            <TableButton
              type="button"
              id="ColTypeSortButton"
              aria-label="Sort by type"
              onClick={() => sortByType(tableData, typeAscending)}
            >
              <span>Type</span>
              {sortKeys[1] ? (
                typeAscending ? (
                  <StyledArrowUpIcon aria-hidden="true" />
                ) : (
                  <StyledArrowDownIcon aria-hidden="true" />
                )
              ) : (
                <React.Fragment />
              )}
            </TableButton>
          </StyledTableHeader>
          <StyledTableHeader
            id="ColDate"
            aria-sort={
              sortKeys[2]
                ? dateAscending
                  ? 'ascending'
                  : 'descending'
                : 'none'
            }
          >
            <TableButton
              type="button"
              id="ColDateSortButton"
              aria-label="Sort by date"
              onClick={() => sortByDate(tableData, dateAscending)}
            >
              <span>Date</span>
              {sortKeys[2] ? (
                dateAscending ? (
                  <StyledArrowUpIcon aria-hidden="true" />
                ) : (
                  <StyledArrowDownIcon aria-hidden="true" />
                )
              ) : (
                <React.Fragment />
              )}
            </TableButton>
          </StyledTableHeader>
          <StyledTableHeader
            id="ColSize"
            aria-sort={
              sortKeys[3]
                ? sizeAscending
                  ? 'ascending'
                  : 'descending'
                : 'none'
            }
          >
            <TableButton
              type="button"
              id="ColSizeSortButton"
              aria-label="Sort by size"
              onClick={() => sortBySize(tableData, sizeAscending)}
            >
              <span>Size</span>
              {sortKeys[3] ? (
                sizeAscending ? (
                  <StyledArrowUpIcon aria-hidden="true" />
                ) : (
                  <StyledArrowDownIcon aria-hidden="true" />
                )
              ) : (
                <React.Fragment />
              )}
            </TableButton>
          </StyledTableHeader>
        </StyledTableHeadRow>
      </StyledTableHead>
      <StyledTableBody>
        {tableData.map((file, key) => dataRow(file, key))}
      </StyledTableBody>
    </StyledTable>
  )
}

export default Table
