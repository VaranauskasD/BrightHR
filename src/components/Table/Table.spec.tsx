import { screen, fireEvent, waitFor } from '@testing-library/react'

import { customRender as render } from '../../utils/test'
import { Table, TableProps } from './Table'

let mockTableProps: TableProps
describe('Table', () => {
  beforeEach(() => {
    mockTableProps = {
      data: [
        {
          'type': 'pdf',
          'name': 'Amplified',
          'added': '2022-02-07',
        },
        {
          'type': 'pdf',
          'name': 'Employee Handbook',
          'added': '2017-01-06',
        },
        {
          'type': 'pdf',
          'name': 'Human Resources Handbook',
          'added': '2014-02-06',
        },
        {
          'type': 'pdf',
          'name': 'Executive Handbook',
          'added': '2012-04-03',
        },
        {
          'type': 'pdf',
          'name': 'Setup Guide',
          'added': '2017-07-02',
        },
        {
          'type': 'pdf',
          'name': 'Statements',
          'added': '2017-02-15',
        },
        {
          'type': 'pdf',
          'name': 'Networks Handbook',
          'added': '2013-02-14',
        },
        {
          'type': 'pdf',
          'name': 'Marketing Handbook',
          'added': '2015-02-06',
        },
        {
          'type': 'pdf',
          'name': 'Public Holiday policy',
          'added': '2016-12-06',
        },
        {
          'type': 'pdf',
          'name': 'Public Holiday policy 2',
          'added': '2016-12-07',
        },
        {
          'type': 'pdf',
          'name': 'Public Holiday policy 3',
          'added': '2016-12-08',
        },
        {
          'type': 'zsv',
          'name': 'Zen',
          'added': '2011-05-16',
        },
        {
          'type': 'pdf',
          'name': 'Public Holiday policy final',
          'added': '2016-12-09',
        },
        {
          'type': 'folder',
          'name': 'Expenses',
          'files': [
            {
              'type': 'doc',
              'name': 'Expenses claim form',
              'added': '2017-05-02',
            },
            {
              'type': 'doc',
              'name': 'Fuel allowances',
              'added': '2017-05-03',
            },
          ],
        },
        {
          'type': 'csv',
          'name': 'Cost centres',
          'added': '2016-08-12',
        },
        {
          'type': 'folder',
          'name': 'Misc',
          'files': [
            {
              'type': 'doc',
              'name': 'Christmas party',
              'added': '2017-12-01',
            },
            {
              'type': 'doc',
              'name': 'Christmas party 2',
              'added': '2019-12-01',
            },
            {
              'type': 'mov',
              'name': 'Welcome to the company!',
              'added': '2015-04-24',
            },
          ],
        },
      ],
      filter: '',
    }
  })

  it('renders correctly', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by name ascending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColNameSortButton'))
    fireEvent.click(screen.getByTestId('ColNameSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Amplified-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by name descending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColNameSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Zen-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by type ascending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColTypeSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Expenses-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by type descending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColNameSortButton'))
    fireEvent.click(screen.getByTestId('ColNameSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Amplified-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by date ascending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColDateSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Zen-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by date descending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColDateSortButton'))
    fireEvent.click(screen.getByTestId('ColDateSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Amplified-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by size ascending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColSizeSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Expenses-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('sorts by size descending', async () => {
    const { asFragment } = render(<Table {...mockTableProps} />)
    fireEvent.click(screen.getByTestId('ColSizeSortButton'))
    fireEvent.click(screen.getByTestId('ColSizeSortButton'))

    let expectedFirstElement = screen.getByTestId('cell-Misc-0')
    let firstElement = screen.getByTestId('tablebody').firstChild?.firstChild
    expect(expectedFirstElement).toEqual(firstElement)
    expect(asFragment()).toMatchSnapshot()
  })
})
