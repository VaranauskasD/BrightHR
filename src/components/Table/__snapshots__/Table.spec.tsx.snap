// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
<DocumentFragment>
  <table
    aria-labelledby="caption"
    class="sc-bczRLJ gmOvJp"
  >
    <caption
      class="sc-gsnTZi jmINyp"
      id="caption"
    >
      Files
    </caption>
    <thead
      class="sc-dkzDqf oORPV"
    >
      <tr
        class="sc-hKMtZM cVghEq"
      >
        <th
          aria-sort="descending"
          class="sc-jSMfEi rZCU"
          id="ColName"
        >
          <button
            aria-label="Sort by name"
            class="sc-gKXOVf swDVD"
            id="ColNameSortButton"
            type="button"
          >
            <span>
              Name
            </span>
            <svg
              aria-hidden="true"
              class="sc-jqUVSM gBWJPk"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M862 465.3h-81c-4.6 0-9 2-12.1 5.5L550 723.1V160c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v563.1L255.1 470.8c-3-3.5-7.4-5.5-12.1-5.5h-81c-6.8 0-10.5 8.1-6 13.2L487.9 861a31.96 31.96 0 0 0 48.3 0L868 478.5c4.5-5.2.8-13.2-6-13.2z"
              />
            </svg>
          </button>
        </th>
        <th
          aria-sort="none"
          class="sc-jSMfEi rZCU"
          id="ColType"
        >
          <button
            aria-label="Sort by type"
            class="sc-gKXOVf swDVD"
            id="ColTypeSortButton"
            type="button"
          >
            <span>
              Type
            </span>
          </button>
        </th>
        <th
          aria-sort="none"
          class="sc-jSMfEi rZCU"
          id="ColDate"
        >
          <button
            aria-label="Sort by date"
            class="sc-gKXOVf swDVD"
            id="ColDateSortButton"
            type="button"
          >
            <span>
              Date
            </span>
          </button>
        </th>
        <th
          aria-sort="none"
          class="sc-jSMfEi rZCU"
          id="ColSize"
        >
          <button
            aria-label="Sort by size"
            class="sc-gKXOVf swDVD"
            id="ColSizeSortButton"
            type="button"
          >
            <span>
              Size
            </span>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      class="sc-iBkjds cTRa-DO"
    >
      <tr
        class="sc-eCYdqJ bmVhaN"
      >
        <td
          class="sc-ftvSup jXKhkW"
        >
          Cost centres
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          csv
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          2016-08-12
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          Missing size
        </td>
      </tr>
      <tr
        class="sc-eCYdqJ hcWJXS"
      >
        <td
          class="sc-ftvSup jXKhkW"
        >
          Employee Handbook
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          pdf
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          2017-01-06
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          Missing size
        </td>
      </tr>
      <tr
        class="sc-eCYdqJ bmVhaN"
      >
        <td
          class="sc-ftvSup jXKhkW"
        >
          Expenses
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          folder
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          Missing Date
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          2
        </td>
      </tr>
      <tr
        class="sc-eCYdqJ hcWJXS"
      >
        <td
          class="sc-ftvSup jXKhkW"
        >
          Misc
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          folder
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          Missing Date
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          2
        </td>
      </tr>
      <tr
        class="sc-eCYdqJ bmVhaN"
      >
        <td
          class="sc-ftvSup jXKhkW"
        >
          Public Holiday policy
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          pdf
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          2016-12-06
        </td>
        <td
          class="sc-ftvSup jXKhkW"
        >
          Missing size
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
